protocol/wire-type-derive/src/lib.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0
+
+#![deny(rustdoc::all)]
+
+//!
+//! This crate contains procedural macro for for deriving WireType trait
+//!
+
+use darling::FromDeriveInput;
+use proc_macro::TokenStream;
+use quote::{format_ident, quote};
+use syn::__private::TokenStream2;
+use syn::{parse_macro_input, Data, DataEnum, DataStruct, DeriveInput, Fields};
+
+#[derive(FromDeriveInput)]
+#[darling(default, attributes(wire_type))]
+struct Opts {
+    version: u16,
+    recurse: usize,
+}
+
+impl std::default::Default for Opts {
+    fn default() -> Self {
+        Opts {
+            version: 1,
+            recurse: 1,
+        }
+    }
+}
+
+#[proc_macro_derive(WireType, attributes(wire_type))]
+pub fn derive(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input);
+    let Opts { version, recurse } = Opts::from_derive_input(&input)
+        .expect("Invalid options for wire_type. Must provide a version number (e.g. #[wire_type(version = 1)]");
+    let shadow = shadow_from_input(input.clone());
+    let DeriveInput {
+        ident,
+        generics: _,
+        data: _,
+        ..
+    } = input;
+
+    let wire_ident = format_ident!("__Wire{}", ident);
+    let shadow_ident = format_ident!("__Shadow{}", ident);
+    let shadow_ident_str = shadow_ident.to_string();
+
+    // this is a recursive way to allow nested versioning
+    // e.g. `{"version":1,"t":{"version":1,"t":{"a":123,"b":321}}}`
+    // which occurs sometimes in the Mina codebase
+    // This may need to be rewritten if it doesn't work in all cases
+    let recurse_attr = if recurse > 1 {
+        let recurse_less_1 = recurse - 1;
+        Some(quote! {
+            #[derive(Clone, WireType)]
+            #[wire_type( recurse = #recurse_less_1 )]
+            #[serde(from = "<Self as WireType>::WireType")]
+            #[serde(into = "<Self as WireType>::WireType")]
+        })
+    } else {
+        None
+    };
+
+    let output = quote! {
+
+        #shadow
```

Comment:
This is referencing the local variable `shadow` which is a copy of the struct/enum that wire type is being implemented on. It is used by `serde_with` to provide the serialization for the type when it is wrapped by the version containing struct.

It breaks the infinite loop which would otherwise exist if you add try to serialize a type as its wire type (which contains the type, which get serialized as the wire type, etc)